[SPARK-18362][SQL] Use TextFileFormat in implementation of CSVFileFormat

JoshRosen opened the pull request:
## What changes were proposed in this pull request?

This patch significantly improves the IO / file listing performance of schema inference in Spark's built-in CSV data source.

Previously, this data source used the legacy `SparkContext.hadoopFile` and `SparkContext.hadoopRDD` methods to read files during its schema inference step, causing huge file-listing bottlenecks on the driver.

This patch refactors this logic to use Spark SQL's `text` data source to read files during this step. The text data source still performs some unnecessary file listing (since in theory we already have resolved the table prior to schema inference and therefore should be able to scan without performing _any_ extra listing), but that listing is much faster and takes place in parallel. In one production workload operating over tens of thousands of files, this change managed to reduce schema inference time from 7 minutes to 2 minutes.

A similar problem also affects the JSON file format and this patch originally fixed that as well, but I've decided to split that change into a separate patch so as not to conflict with changes in another JSON PR.

## How was this patch tested?

Existing unit tests, plus manual benchmarking on a production workload.

JoshRosen committed:
Use text data source in CSV and JSON data sources.

JoshRosen committed:
Don't check file existence.

JoshRosen committed:
Fix name.

JoshRosen committed:
Actually use `lines`

JoshRosen committed:
Clean up CSV code to reduce number of scans.

SparkQA commented:
**[Test build #68359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68359/consoleFull)** for PR 15813 at commit [`acce60d`](https://github.com/apache/spark/commit/acce60d4f584f549a44f6dd90e3db5e6694a61f0).

SparkQA commented:
**[Test build #68359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68359/consoleFull)** for PR 15813 at commit [`acce60d`](https://github.com/apache/spark/commit/acce60d4f584f549a44f6dd90e3db5e6694a61f0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68359/
Test FAILed.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala:
```diff
@@ -210,14 +196,21 @@ class CSVFileFormat extends TextBasedFileFormat with DataSourceRegister {
+          paths = inputPaths,
+          className = classOf[TextFileFormat].getName
+        ).resolveRelation(checkFilesExist = false))
+        .select("value").as[String](Encoders.STRING)
```
I copied this logic from the `text` method in `DataFrameReader`, so that's where the `value` came from.

JoshRosen commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #68432 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68432/consoleFull)** for PR 15813 at commit [`acce60d`](https://github.com/apache/spark/commit/acce60d4f584f549a44f6dd90e3db5e6694a61f0).

SparkQA commented:
**[Test build #68432 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68432/consoleFull)** for PR 15813 at commit [`acce60d`](https://github.com/apache/spark/commit/acce60d4f584f549a44f6dd90e3db5e6694a61f0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68432/
Test FAILed.

JoshRosen commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #68674 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68674/consoleFull)** for PR 15813 at commit [`acce60d`](https://github.com/apache/spark/commit/acce60d4f584f549a44f6dd90e3db5e6694a61f0).

SparkQA commented:
**[Test build #68674 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68674/consoleFull)** for PR 15813 at commit [`acce60d`](https://github.com/apache/spark/commit/acce60d4f584f549a44f6dd90e3db5e6694a61f0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68674/
Test FAILed.

JoshRosen committed:
Ensure that inferSchema is not called with empty set of files.

SparkQA commented:
**[Test build #68800 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68800/consoleFull)** for PR 15813 at commit [`3082844`](https://github.com/apache/spark/commit/3082844686aa8344d5fc350c790d85aa4f8ba500).

SparkQA commented:
**[Test build #68800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68800/consoleFull)** for PR 15813 at commit [`3082844`](https://github.com/apache/spark/commit/3082844686aa8344d5fc350c790d85aa4f8ba500).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68800/
Test PASSed.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala:
```diff
@@ -173,35 +178,17 @@ class CSVFileFormat extends TextBasedFileFormat with DataSourceRegister {
     if (options.isCommentSet) {
       val comment = options.comment.toString
-      rdd.filter { line =>
+      lines.filter { line =>
```
Using untyped `filter` can be more performant here since we don't need to pay for the extra de/serialization costs:

```scala
lines.filter(length(trim($"value")) > 0 && $"value".startsWith(comment))
```

NathanHowell commented:
Any thoughts on modifying `JsonToStruct` to support arrays (and options), then parsing could be something like:

```
dataset.select(
  Column(Inline(
    JsonToValue(
      ArrayType(schema),
      options,
      Column("value").expr))))
```

Likely out of scope for this pull request, but if there is a push to migrate from `RDD[T]` to `Dataset[T]` it would clean things up a bit.

JoshRosen committed:
Use SQL filter to speed up first line query.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into use-text-data-source-in-csv-and-json

JoshRosen committed:
Undo JSON changes.

SparkQA commented:
**[Test build #69418 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69418/consoleFull)** for PR 15813 at commit [`b01a307`](https://github.com/apache/spark/commit/b01a307499a5b2a00039cb4b195d05a40157438c).

SparkQA commented:
**[Test build #69418 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69418/consoleFull)** for PR 15813 at commit [`b01a307`](https://github.com/apache/spark/commit/b01a307499a5b2a00039cb4b195d05a40157438c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69418/
Test PASSed.

JoshRosen commented:
@NathanHowell, I've gone ahead and removed the JSON changes from this PR; now it only touches CSV and thus should not conflict with your work.

@liancheng, want to give this a final review? I've addressed your earlier comments.

rxin commented:
Merging in master.

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala:
```diff
@@ -173,51 +179,37 @@ class CSVFileFormat extends TextBasedFileFormat with DataSourceRegister {
       val charset = options.charset
-      sparkSession.sparkContext
-        .hadoopFile[LongWritable, Text, TextInputFormat](location)
+      val rdd = sparkSession.sparkContext
```
@JoshRosen do you know why the special handling for non-utf8 encoding is needed? I would think TextFileFormat itself already supports that since it is reading it in from Hadoop Text.

HyukjinKwon replied:
@rxin, I made a PR to address it at https://github.com/apache/spark/pull/29063 FYI.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala:
```diff
@@ -173,51 +179,37 @@ class CSVFileFormat extends TextBasedFileFormat with DataSourceRegister {
       val charset = options.charset
-      sparkSession.sparkContext
-        .hadoopFile[LongWritable, Text, TextInputFormat](location)
+      val rdd = sparkSession.sparkContext
```
I'm not sure; I think this was a carryover from `spark-csv`.

rxin replied:
cc @falaki 
Can you chime in?
